File: helpers/update-sources.modules/ContentDispositionParameterKey.rb
=begin
 
 ContentDispositionParameterKey.rb
   ©︎ 2017 YOCKOW.
     Licensed under MIT License.
     See "LICENSE.txt" for more information.
 
=end

require File.realpath('./_common_.rb', File.dirname(__FILE__))
require File.realpath('./_swift_common_.rb', File.dirname(__FILE__))

module ContentDispositionParameterKey
  PATH = 'Sources/CGIResponder/ContentDispositionParameterKey.swift'
  URLs = [
    URI.parse('https://www.iana.org/assignments/cont-disp/cont-disp-2.csv')
  ]
  
  extend SWIFT_COMMON
  
  def write(file)
    keys = []
    URLs.each {|url|
      $stdout.puts("- Fetching #{url}")
      url.parse_csv {|row|
        next if row[0] !~ /^[\-a-z]+$/
        keys.push(row[0])
      }
    }
    raise "No data about Content Disposition Parameter Keys" if keys.count < 1
    
    file.puts(head(PATH, URLs))
    file.puts('public struct ContentDispositionParameterKey: RawRepresentable {')
    file.puts('  public let rawValue: String')
    file.puts('  public init(rawValue:String) { self.rawValue = rawValue }')
    file.puts('}')
    
    file.puts('extension ContentDispositionParameterKey: Hashable {')
    file.puts('  public var hashValue: Int { return self.rawValue.hashValue }')
    file.puts('  public static func ==(lhs:ContentDispositionParameterKey, rhs:ContentDispositionParameterKey) -> Bool {')
    file.puts('    return lhs.rawValue == rhs.rawValue')
    file.puts('  }')
    file.puts('}')
    
    file.puts('extension ContentDispositionParameterKey {')
    keys.each {|key|
      keyword = key.to_keyword
      file.write('  public static let ')
      file.write((keyword.reserved?) ? "`#{keyword}`" : "#{keyword}")
      file.write(" = ContentDispositionParameterKey(rawValue:\"#{key}\")")
      file.write("\n")
    }
    file.puts('}')
  end
  module_function :write
end

File: Sources/Supporters/HTTP/ContentDispositionParameterKey.swift
/* *************************************************************************************************
 ContentDispositionParameterKey.swift
   © 2018-2019 YOCKOW.
     Licensed under MIT License.
     See "LICENSE.txt" for more information.
 ************************************************************************************************ */

/// Represents the key for parameter of Content Disposition.
/// It is almost always "filename"
public struct ContentDispositionParameterKey: RawRepresentable {
  public typealias RawValue = String
  public let rawValue: String
  public init(rawValue:String) {
    self.rawValue = rawValue
  }
}

extension ContentDispositionParameterKey: Hashable {
  public static func ==(lhs:ContentDispositionParameterKey, rhs:ContentDispositionParameterKey) -> Bool {
    return lhs.rawValue == rhs.rawValue
  }
  
  public func hash(into hasher:inout Hasher) {
    hasher.combine(self.rawValue)
  }
}

extension ContentDispositionParameterKey: ExpressibleByStringLiteral {
  public init(stringLiteral string:String) {
    self.init(rawValue:string)
  }
}
